packages/shared/invokeGuardedCallbackImpl.js

```diff
@@ -7,79 +7,42 @@
-  ) {
-    const fakeNode = document.createElement('react');
-
-    invokeGuardedCallbackImpl = function invokeGuardedCallbackDev<
-      Args: Array<mixed>,
-      Context,
-      // $FlowFixMe[missing-this-annot]
-    >(name: string | null, func: (...Args) => mixed, context: Context): void {
-      // If document doesn't exist we know for sure we will crash in this method
-      // when we call document.createEvent(). However this can cause confusing
-      // errors: https://github.com/facebook/create-react-app/issues/3482
-      // So we preemptively throw with a better message instead.
-      if (typeof document === 'undefined' || document === null) {
-        throw new Error(
-          'The `document` global was defined when React was initialized, but is not ' +
-            'defined anymore. This can happen in a test environment if a component ' +
-            'schedules an update from an asynchronous callback, but the test has already ' +
-            'finished running. To solve this, you can either unmount the component at ' +
-            'the end of your test (and ensure that any asynchronous operations get ' +
-            'canceled in `componentWillUnmount`), or you can change the test itself ' +
-            'to be asynchronous.',
-        );
-      }
+  if (__DEV__) {
+    // In DEV mode, we use a special version
+    // that plays more nicely with the browser's DevTools. The idea is to preserve
+    // "Pause on exceptions" behavior. Because React wraps all user-provided
+    // functions in invokeGuardedCallback, and the production version of
+    // invokeGuardedCallback uses a try-catch, all user exceptions are treated
+    // like caught exceptions, and the DevTools won't pause unless the developer
+    // takes the extra step of enabling pause on caught exceptions. This is
+    // unintuitive, though, because even though React has caught the error, from
+    // the developer's perspective, the error is uncaught.
+    //
+    // To preserve the expected "Pause on exceptions" behavior, we don't use a
+    // try-catch in DEV. Instead, we synchronously dispatch a fake event to a fake
+    // DOM node, and call the user-provided callback from inside an event handler
+    // for that fake event. If the callback throws, the error is "captured" using
+    // a global event handler. But because the error happens in a different
+    // event loop context, it does not interrupt the normal program flow.
+    // Effectively, this gives us try-catch behavior without actually using
+    // try-catch. Neat!
+
+    // Check that the browser supports the APIs we need to implement our special
+    // DEV version of invokeGuardedCallback
+    if (
+      typeof window !== 'undefined' &&
+      typeof window.dispatchEvent === 'function' &&
+      typeof document !== 'undefined' &&
+      // $FlowFixMe[method-unbinding]
+      typeof document.createEvent === 'function'
+    ) {
+      const fakeNode = document.createElement('react');
```

Comment:
This will create a new element thousands of times in DEV mode. We need to make DEV mode much faster, not slower.

Why do we need the latest window/document though? Even if we change it later, we could still dispatch against the original window/document.

Couldn't we just stash `const win = window, doc = document;` in the module and then always use the same fake window/document for creating the event and that it gets dispatched to?

Reply:
I may misunderstand everywhere this is used but we're talking about once (or twice) per error right?

No reason not to change it but your comment makes me think I misunderstand something about how this is used

Reply:
No this is called everywhere there might be an error we don't want to affect other branches even if there isn't one. It's like React's try/catch.

Reply:
Hm... Looks like we've cut down usage since we have the double render in DEV for rendering the first one doesn't use this and allows for suspense.

We also stopped using it in the commit phase. https://github.com/facebook/react/pull/21666

So I guess now it's no longer used that way in most cases. It's still used that way for events etc. but that probably indicates that the whole thing should be refactored since it's not used as initially intended anymore.